Limit orphan pool size by total number of txns as well as by total pool bytes

ptschip opened the pull request:
Keep track of the orphan pool bytes so we can limit by bytes as well as the limiting by number of txns, whichever limit is hit first. (We still want to keep the limiting by number of txns in place because that keeps the bloom filters used in xthins from getting too large.)

ptschip committed:
Keep track of the in memory orphan tx size

Start tracking this so we can limit the orphan cache size by
total bytes as well as by number of transactions.

Limit size of orphan pool by total bytes as well as transactions

Update code comments to reflect the new orphan pool handling.

Add unit tests for orphan pool max size allowed

ptschip committed:
Remove dead orphan pool code

For a long time we have not used this code for removing orphans by
peer and rely instead on number of orphans in the pool.

ptschip commented:
Taking this out of WIP.

There is a new unit test for this but you can also test this manually by turning on debug=mempool and manually editing main.cpp line 5650 and make nMaxOrphanPoolSize = 50000

ptschip commented:
rebased

ptschip commented:
rebased again

sickpig reviewed (approved):
compile and tested (c++ and py suites) 
all tests pass, even the new one in `DoS_tests` that check the functionality of `LimitOrphanTxSize` both  for max number of txs and max orphans pool size.

gandrewstone commented on src/main.cpp:
```diff
@@ -796,35 +794,26 @@ bool AddOrphanTx(const CTransaction &tx, NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(c
     if (sz > MAX_STANDARD_TX_SIZE)
     {
         LogPrint("mempool", "ignoring large orphan tx (size: %u, hash: %s)\n", sz, hash.ToString());
+        printf("ignoring large orphan tx (size: %u, hash: %s)\n", sz, hash.ToString().c_str());
```
accidental debug print left in?

ptschip replied:
oops , done.

ptschip committed:
take out accidental printf left in code

gandrewstone commented:
Please search through the code for every use of mapOrphanTransactions and ensure that there is no other place where it is modified without changing your new counter.  
For example, it looks like it is cleared at main.cpp:4906 but the counter is not cleared.

sickpig commented:
Another place in `main.cpp` where the map is cleared:

```
$> grep -wnr mapOrphanTransactions * | grep clear 
main.cpp:4906:        mapOrphanTransactions.clear();
main.cpp:7657:        mapOrphanTransactions.clear();
```

ptschip committed:
Set nBytesOrphanPool to zero on startup

This happens during initialization when UnloadBlockIndex() is called
during startup.

ptschip commented:
@sickpig @gandrewstone looks like there are two places as @sickpig has shown, but only line 4906 is relevant which gets called during startup during UnloadBlockIndex()...the other place line7657 happens during shutdown so I don' think we need it there.

pushing up now.

sickpig commented:
@ptschip 
I wonder if we should use a wrapper in such a way that every time we clear the map we are also going to reset the counter? 

That will defend us from possible future problems, e.g. if we need  `mapOrphanTransactions.clear()` to any other codepath...

ptschip commented:
@sickpig not sure it's worth it, the only time we will ever clear the map is on exit or startup.

ptschip commented:
@sickpig also if we created a wrapper there is nothing preventing anyone from not using it, but you bring up a good point and I'm thinking we should just put a check for map size at the top of addorphantx...so if the map got cleared somewhere we can set the bytes to zero before adding the first orphan.

ptschip committed:
Zero the nBytesOrphanPool if the orphan map has been cleared.

This is done before we add an orphantx in the event that the map
was cleared somewhere.

gandrewstone commented:
When dealing with software state coherence issues, I would like to see DbgAssert() used to cause the code to core dump in debug builds and correct in release builds.  Otherwise, correcting state coherence can mask bugs that ought to be fixed.  Specifically:

if (mapOrphanTransactions.empty())
{
    DbgAssert(nBytesOrphanPool==0, nBytesOrphanPool = 0);
}

This is not important enough to block this PR so merging.  Please add the above to another PR when you get the chance.
